joomla/joomla-cms
# [4.0] Enforce2fa

### Summary of Changes
Added Enforce Two Factor Authentication to the core.

This change is supposed to force the user both in Back- and or Frontend to first setup their Two Factor Authentication before being able to proceed.
The enforcement triggers when the user tries to navigate somewhere else or the page is being reloaded.

### Testing Instructions
1. Setup Joomla! 4.0-dev build
2. In the menu go the System 
3. Click on Plugins
4. Then search for Two Factor Authentication 
5. Enable a Two Factor Authentication Plugin of your choice

![image](https://user-images.githubusercontent.com/54268350/64693046-b5b17680-d496-11e9-9285-9236485d4d57.png)


6. Edit the Two Factor Authentication Methode that you choose
7. Select where the Authentication should be enabled Front- or Backend or Both (default is both Front- and Backend)
8. After that click save & close
9. In the menu now go again to System
10. Click on Global Configuration
11. Now scrolle down and click on Users at the bottom left
12. At the bottom there should be a label Enforce Two Factor Authentication
13. Select now where and if Two Factor Authentication should be enforced, note it should be the same setting as in point 8

![image](https://user-images.githubusercontent.com/54268350/64767745-8c512300-d548-11e9-8bc6-575068311208.png)


14. After that click save & close.
15. Now depending on your settings go to the Frontend and log in or try to navigate somewhere else.

### Expected result
You will be redirected to your user edit page until you finished successfully setting up Two Factor Authentication.

### Documentation Changes Required
Description of the feature in the documentation, hence it is a core feature
